lib/betterdoc/containerservice/helpers/link_helper.rb

```diff
@@ -0,0 +1,45 @@
+require 'cgi'
+
+module Betterdoc
+  module Containerservice
+    module Helpers
+      module LinkHelper
+        extend ActiveSupport::Concern
+
+        # rubocop:disable Metrics/CyclomaticComplexity,Metrics/PerceivedComplexity
+        def create_stacker_link(target_path, parameters = {})
+
+          base_url_from_request = resolve_stacker_base_url_from_request
+          base_url_from_environment = resolve_stacker_base_url_from_environment
+          base_url = base_url_from_request if base_url_from_request.present?
+          base_url = base_url_from_environment if base_url.blank? && base_url_from_environment.present?
+          base_url = '/' if base_url.blank?
+
+          result_url = base_url.dup
+          result_url << '/' unless base_url.end_with?('/') || target_path.start_with?('/')
+          result_url << target_path unless result_url.end_with?('/') && target_path.start_with?('/')
+          result_url << target_path[1, target_path.length] if result_url.end_with?('/') && target_path.start_with?('/')
+
+          parameters.each_with_index do |(parameter_name, parameter_value), parameter_index|
+            result_url << (parameter_index.zero? && !result_url.include?('?') ? '?' : '&')
+            result_url << CGI.escape(parameter_name.to_s) << '=' << CGI.escape(parameter_value.to_s)
+          end
+          result_url
+
+        end
+        # rubocop:enable Metrics/CyclomaticComplexity,Metrics/PerceivedComplexity
```

Comment:
Just a quick idea how to refactor this:

```ruby
def create_stacker_link(target_path, parameters = {})
  base_url = request.headers['HTTP_X-STACKER-ROOT-URL'].presence || ENV['STACKER_ROOT_URL'].presence || "/"
  [URI.join(base_url, target_path), parameters.to_query].select(&:present?).join("?")  
end
```

This is just a quick idea, you can further improve and/or play with https://docs.ruby-lang.org/en/2.1.0/URI.html if you like. `to_query` is Rails method https://api.rubyonrails.org/classes/Hash.html#method-i-to_query

Also, since it returns url I would maybe rename this method to something more rails like `stacker_url_for(...)` and then add other method called `stacker_link_to(text, url, options)` that will delegate to standard `link_to` helper but will pregenarate url.

```ruby
def stacker_link_to(text, url, html_options = {})
  link_to(text, stacker_url_for(url), html_options)
end
``` 

Only problem is how to pass parameters in that case. Could be maybe done like this.
```ruby
def stacker_link_to(text, url, html_options = {})
  parameters = html_options.delete(:parameters)
  link_to(text, stacker_url_for(url, parameters), html_options)
end
```

Reply:
Good idea, will take a look at this!